xarray/core/variable.py

```diff
@@ -376,29 +377,89 @@ def _item_key_to_tuple(self, key):
         else:
             return key
 
+    def _broadcast_indexes(self, key):
+        """
+        Parameters
+        -----------
+        key: One of
+            array
+            a mapping of dimension names to index.
+
+        Returns
+        -------
+        dims: Tuple of strings.
+            Dimension of the resultant variable.
+        indexers: list of integer, array-like, or slice. This is aligned
+            along self.dims.
+        """
+        key = self._item_key_to_tuple(key)  # key is a tuple
+        # key is a tuple of full size
+        key = indexing.expanded_indexer(key, self.ndim)
+        basic_indexing_types = integer_types + (slice,)
+        if all([isinstance(k, basic_indexing_types) for k in key]):
+            return self._broadcast_indexes_basic(key)
+        else:
+            return self._broadcast_indexes_advanced(key)
+
+    def _broadcast_indexes_basic(self, key):
+        dims = tuple(dim for k, dim in zip(key, self.dims)
+                     if not isinstance(k, integer_types))
+        return dims, key
+
+    def nonzero(self):
+        """ Equivalent numpy's nonzero but returns a tuple of Varibles. """
+        if isinstance(self._data, (np.ndarray, pd.Index, PandasIndexAdapter)):
+            nonzeros = np.nonzero(self._data)
+        elif isinstance(self._data, dask_array_type):
+            # TODO we should replace dask's native nonzero
+            # after https://github.com/dask/dask/issues/1076 is implemented.
+            nonzeros = np.nonzero(self.load()._data)
+
+        return tuple([as_variable(nz, name=dim) for nz, dim
+                      in zip(nonzeros, self.dims)])
+
+    def _isbool_type(self):
+        """ Return if the variabe is bool or not """
+        if isinstance(self._data, (np.ndarray, PandasIndexAdapter, pd.Index)):
+            return self._data.dtype is np.dtype('bool')
+        elif isinstance(self._data, dask_array_type):
+            raise NotImplementedError
+
+    def _broadcast_indexes_advanced(self, key):
+        variables = []
+
+        for dim, value in zip(self.dims, key):
+            if isinstance(value, slice):
+                value = np.arange(self.sizes[dim])[value]
```

Comment:
Use the `slice.indices` method here, to construct the desired array without indexing:
```
In [14]: x = slice(0, 10, 2)

In [15]: np.arange(*x.indices(5))
Out[15]: array([0, 2, 4])
```